README

Laser2Wav: a software-only decoder for the physical signal as present on a CD.


python/HISTORY

This file documents the history of the laser2wav package.

--- version 0.0.2:

* Initial release

--- version 0.0.5:

* fixed silly "to_signed" bug
* added HISTORY
* renamed source and data files
* added lots of comments
* cleaned up "galois field" code
* emit a PCM-encoded WAV file instead of a text file
* laser input is now assumed to be the raw signal
  (not the delta-signal as before).

--- 23 May 2015:

* transferring the repository and its data to Github.


python/laser2wav.py
#! /usr/bin/env python3

# laser2wav.py
#
# Copyright (c) 2005--2015 by Sidney Cadot <sidney@jigsaw.nl>
# This software is licensed under the GNU General Public License (GPL).
#
# This file is part of Laser2Wav, a software-only implementation of
# an audio CD decoder.


import sys, wave
import galois_field as gf
from efm import EFM

def analyze_frame(frame):
    assert len(frame) == 588 # verify that each frame is 588 bits long
    sync24 = frame[0:24]
    merge3 = frame[24:27] # ignore
    assert sync24 == "100000000001000000000010"

    framedata = []
    for i in range(33):
        channel14 = frame[27+17*i:41+17*i]
        merge3    = frame[41+17*i:44+17*i] # ignore
        data8 = EFM[channel14]
        framedata.append(data8)

    return (framedata[0], framedata[1:])

def invert_last_16(Q):
    return Q[:-16] + str.join("", [str(1 - int(x)) for x in Q[-16:]])

def valid_crc(Q):
    # this is a simple way to check the CRC
    if len(Q) == 0:
        return True
    if Q[0] == '0':
        return valid_crc(Q[1:])
    if len(Q)<17:
        return False

    # subtract multiple of the CRC generator polynomial
    Q = [int(x) for x in Q]
    Q[ 0] = 1 - Q[ 0]
    Q[ 4] = 1 - Q[ 4]
    Q[11] = 1 - Q[11]
    Q[16] = 1 - Q[16]
    Q = str.join("", [str(x) for x in Q])
    return valid_crc(Q)

def to_signed(n):
    if n >= 32768:
        n -= 65536
    return n

def to_unsigned(n):
    if n < 0:
        n += 65536
    return n

def bcd2int(S):
    assert len(S)==8
    hi = int(S[0:4], 2)
    #assert (hi>=0) and (hi<=9)
    lo = int(S[4:8], 2)
    #assert (lo>=0) and (lo<=9)
    return hi*10+lo

def analyze_control_stream(control_stream):

    print("Analyzing control stream ...")

    all_zeros = 96 * "0"
    all_ones  = 96 * "1"
    ok_map = {False: "ERROR", True: "ok"}

    while len(control_stream) >= 98:
        if control_stream[0] != 'SYNC0':
            # discard a control byte - not part of a sector
            control_stream = control_stream[1:]
        else:
            assert control_stream[0] == 'SYNC0'
            assert control_stream[1] == 'SYNC1'

            control_sector = control_stream[2:98]
            control_stream = control_stream[98:]

            P = str.join("", [str(int((x & 0x80) != 0)) for x in control_sector])
            Q = str.join("", [str(int((x & 0x40) != 0)) for x in control_sector])
            R = str.join("", [str(int((x & 0x20) != 0)) for x in control_sector])
            S = str.join("", [str(int((x & 0x10) != 0)) for x in control_sector])
            T = str.join("", [str(int((x & 0x08) != 0)) for x in control_sector])
            U = str.join("", [str(int((x & 0x04) != 0)) for x in control_sector])
            V = str.join("", [str(int((x & 0x02) != 0)) for x in control_sector])
            W = str.join("", [str(int((x & 0x01) != 0)) for x in control_sector])

            p_ok = (P in [all_zeros, all_ones])
            q_ok = valid_crc(invert_last_16(Q))
            r_ok = (R == all_zeros) # for most CDs
            s_ok = (S == all_zeros) # for most CDs
            t_ok = (T == all_zeros) # for most CDs
            u_ok = (U == all_zeros) # for most CDs
            v_ok = (V == all_zeros) # for most CDs
            w_ok = (W == all_zeros) # for most CDs

            print("    Sector sub-channels:")
            print("        P: <{}> {}".format(P, ok_map[p_ok]))
            print("        Q: <{}> {}".format(Q, ok_map[q_ok]))

            if q_ok:
                q_control = Q[0:4]
                q_mode    = int(Q[4:8], 2)
                q_data    = Q[8:80]
                if q_mode == 1:
                    (tno,index,min,sec,frac,zero,amin,asec,afrac) = [bcd2int(q_data[8*i:8*i+8]) for i in range(9)]
                    assert zero == 0
                    print("              CONTROL .......... : <{}> [#channels / reserved / copy-protect / pre-emphasis]".format(q_control))
                    print("              MODE ............. : {}".format(q_mode))
                    print("              DATA ............. : trackno. {:02d} indexno. {:02d} track-time {:02d}:{:02d}.{:02d} [mm:ss.ff] disc-time {:02d}:{:02d}.{:02d} [mm:ss.ff]".format(tno, index, min,  sec,  frac, amin, asec, afrac))
                else:
                    print("              CONTROL .......... : <{}>".format(q_control))
                    print("              MODE ............. : {}".format(q_mode))
                    print("              DATA ............. : <{}>".format(q_data))

            print("        R: <{}> {}".format(R, ok_map[r_ok]))
            print("        S: <{}> {}".format(S, ok_map[s_ok]))
            print("        T: <{}> {}".format(T, ok_map[t_ok]))
            print("        U: <{}> {}".format(U, ok_map[u_ok]))
            print("        V: <{}> {}".format(V, ok_map[v_ok]))
            print("        W: <{}> {}".format(W, ok_map[w_ok]))

def verify_data_stream(data):

    errorsP = 0
    errorsQ = 0

    # make the Reed-Solomon check calculations

    invert = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 255, 255, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 255, 255]

    # check C1 a.k.a. P-parity
    print("Checking C1 / P parity ...")
    p_delay = [0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1]
    for i in range(1, len(data)):
        z_list = []
        for hp_row in range(4):
            z = gf.zero
            for j in range(32):
                z = gf.add(z, gf.multiply(gf.power(gf.alpha, hp_row*(31-j)), data[i-p_delay[j]][j]^invert[j]))
            z_list.append(z)    
        if z_list != [gf.zero, gf.zero, gf.zero, gf.zero]:
            print("P PARITY: ERROR DETECTED IN FRAME # {}".format(i))
            errorsP = errorsP + 1

    # check C2 a.k.a. Q-parity
    print("Checking C2 / Q parity ...")
    q_delay = [107, 104, 99, 96, 91, 88, 83, 80, 75, 72, 67, 64, 59, 56, 51, 48, 43, 40, 35, 32, 27, 24, 19, 16, 11, 8, 3, 0]
    for i in range(107, len(data)):
        z_list = []
        for hq_row in range(4):
            z = gf.zero
            for j in range(28):
                z  = gf.add(z, gf.multiply(gf.power(gf.alpha, hq_row*(27-j)), data[i-q_delay[j]][j]^invert[j]))
            z_list.append(z)
        if z_list!= [gf.zero, gf.zero, gf.zero, gf.zero]:
            print("Q PARITY: ERROR DETECTED IN FRAME # {}".format(i))
            errorsQ = errorsQ + 1

    return (errorsP, errorsQ)

def extract_audio_stream(data):

    audio_data = []

    for i in range(105, len(data)):

        L0  = to_signed(data[i-105][ 0]*256 + data[i-102][ 1])
        L2  = to_signed(data[i- 97][ 2]*256 + data[i- 94][ 3])
        L4  = to_signed(data[i- 89][ 4]*256 + data[i- 86][ 5])

        R0  = to_signed(data[i- 81][ 6]*256 + data[i- 78][ 7])
        R2  = to_signed(data[i- 73][ 8]*256 + data[i- 70][ 9])
        R4  = to_signed(data[i- 65][10]*256 + data[i- 62][11])

        L1  = to_signed(data[i- 43][16]*256 + data[i- 40][17])
        L3  = to_signed(data[i- 35][18]*256 + data[i- 32][19])
        L5  = to_signed(data[i- 27][20]*256 + data[i- 24][21])

        R1  = to_signed(data[i- 19][22]*256 + data[i- 16][23])
        R3  = to_signed(data[i- 11][24]*256 + data[i-  8][25])
        R5  = to_signed(data[i-  3][26]*256 + data[i-  0][27])

        audio_data.extend([(L0, R0), (L1, R1), (L2, R2), (L3, R3), (L4, R4), (L5, R5)])

    return audio_data

def write_wav_file(filename_out, audio_data):

    # Python's WAV-file writer expects raw data, encoded as a string.

    wave_data = bytearray()
    for (left, right) in audio_data:
        left = to_unsigned(left)
        right = to_unsigned(right)
        wave_data.append(left % 256)
        wave_data.append(left // 256)
        wave_data.append(right % 256)
        wave_data.append(right // 256)

    # Write the WAV file.

    wf = wave.open(filename_out, "w")
    wf.setnchannels(2) # stereo
    wf.setsampwidth(2) # 16-bit signal; 2 bytes per sample
    wf.setframerate(44100)
    wf.writeframes(wave_data)
    wf.close()

def main():

    if len(sys.argv) != 3:
        print("Usage: python laser2wav.py <laser-in.dat> <waveform-out.wav>")
        return

    filename_in  = sys.argv[1]
    filename_out = sys.argv[2]

    print("Reading raw signal ...")
    with open(filename_in, "rb") as f:
        raw_signal = f.read()

    print("Converting to delta signal...")
    raw_signal = [int (x) for x in raw_signal]
    delta_signal = [str(raw_signal[i] ^ raw_signal[i - 1]) for i in range(1, len(raw_signal))]
    delta_signal = str.join("", delta_signal)

    print("Walking frames...")
    control_stream = []
    data_stream    = []

    z = 0
    while True:
        z = delta_signal.find("100000000001000000000010", z)
        if z < 0:
            break

        frame = delta_signal[z:z+588]
        if len(frame) == 588:
            (control, data) = analyze_frame(frame)
            control_stream.append(control)
            data_stream.append(data)
        z = z + 588

    analyze_control_stream(control_stream)

    verify_data_stream(data_stream)

    print("Extracting audio data ...")
    audio_data = extract_audio_stream(data_stream)

    print("Writing WAV file ...")
    write_wav_file(filename_out, audio_data)

if __name__ == "__main__":
    main()


python/wav_format.py
# wav_format.py
#
# Copyright (c) 2005--2015 by Sidney Cadot <sidney@jigsaw.nl>
# This software is licensed under the GNU General Public License (GPL).
#
# This file is part of laser2wav, a software-only implementation of
# an audio CD decoder.


def little_endian_long(n):
    return [(n) & 255, (n>>8) & 255, (n>>16) & 255, (n>>24) & 255]

def little_endian_short(n):
    return [(n) & 255, (n>>8) & 255]

def clip(n):
    if n < -32768:
        n = -32768
    if n>32767:
        n = 32767
    return n

def to_unsigned(n):
    if n < 0:
        n += 65536
    return n

def wave_file_string(audio_samples):

    bytes = []

    num_samples     = len(audio_samples)
    subchunk1size   = 16
    audio_format    = 1
    num_channels    = 2
    sample_rate     = 44100
    bits_per_sample = 16
    byte_rate       = sample_rate  * num_channels * bits_per_sample/8
    block_align     = num_channels * bits_per_sample/8
    subchunk2size   = num_samples  * num_channels * bits_per_sample/8

    chunksize       = 4 + (8+subchunk1size) + (8+subchunk2size)

    bytes.extend([ord(x) for x in "RIFF"])
    bytes.extend(little_endian_long(chunksize))
    bytes.extend([ord(x) for x in "WAVE"])
    bytes.extend([ord(x) for x in "fmt "])
    bytes.extend(little_endian_long(subchunk1size))
    bytes.extend(little_endian_short(audio_format))
    bytes.extend(little_endian_short(num_channels))
    bytes.extend(little_endian_long(sample_rate))
    bytes.extend(little_endian_long(byte_rate))
    bytes.extend(little_endian_short(block_align))
    bytes.extend(little_endian_short(bits_per_sample))
    bytes.extend([ord(x) for x in "data"])
    bytes.extend(little_endian_long(subchunk2size))

    for (L, R) in audio_samples:
        bytes.extend(little_endian_short(to_unsigned(clip(L))))
        bytes.extend(little_endian_short(to_unsigned(clip(R))))

    return str.join("", map(chr, bytes))


make-images/make-signal-diagram.py
#! /usr/bin/env python

import Image, ImageDraw

def plot(im, y, x, fillcolor):
    draw = ImageDraw.Draw(im)
    draw.rectangle(((17*x+0, 3*y+0), (17*x+17, 3*y+3)), outline = (100,100,100), fill = fillcolor)

def plot2(im, y, x, fillcolor):
    draw = ImageDraw.Draw(im)
    draw.rectangle(((17*x+0, 3*y+0), (17*x+17, 3*y+3)), outline = (0,0,0), fill = fillcolor)

def plot_q(im, y):
    q_delay = [107, 104, 99, 96, 91, 88, 83, 80, 75, 72, 67, 64, 59, 56, 51, 48, 43, 40, 35, 32, 27, 24, 19, 16, 11, 8, 3, 0]
    for x in range(28):
        plot2(im, y - q_delay[x], 1 + x, (0, 255, 0))

def plot_p(im, y):
    p_delay = [0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1]
    for x in range(32):
        plot2(im, y - p_delay[x], 1 + x, (255, 255, 0))

def plot_sound(im, y):
    audio_bytes = [
        (105,  0), (102,  1),
        ( 97,  2), ( 94,  3),
        ( 89,  4), ( 86,  5),
        ( 81,  6), ( 78,  7),
        ( 73,  8), ( 70,  9),
        ( 65, 10), ( 62, 11),
        ( 43, 16), ( 40, 17),
        ( 35, 18), ( 32, 19),
        ( 27, 20), ( 24, 21),
        ( 19, 22), ( 16, 23),
        ( 11, 24), (  8, 25),
        (  3, 26), (  0, 27)
    ]

    for (oy, x) in audio_bytes:
        plot2(im, y - oy, 1 + x, (255, 0, 0))

im = Image.new('RGB', (562, 3 * 3 * 98 + 1))

for x in range(33):
    for y in range(3 * 98):

        if x in [0]:
            if y % 98 in [0, 1]:
                fillcolor = (204, 153, 204)
            else:
                fillcolor = (255, 204, 255)
        elif x in [1, 3, 5, 7,  9, 11, 17, 19, 21, 23, 25, 27]:
            fillcolor = (230, 230, 255)
        elif x in [2, 4, 6, 8, 10, 12, 18, 20, 22, 24, 26, 28]:
            fillcolor = (220, 220, 255)
        elif x in [13, 14, 15, 16]:
            fillcolor = (204, 255, 204)
        elif x in [29, 30, 31, 32]:
            fillcolor = (255, 255, 204)
        else:
            fillcolor = 'white'

        plot(im, y, x, fillcolor)

plot_p(im, 98 - 20)
plot_q(im, 205 - 10)
plot_sound(im, 203)

im.save('signal-diagram.png')


make-images/make-laserbits-images.py
#! /usr/bin/env python

import Image

with open("laserbits.txt") as f:
    raw_signal = f.read()

# The raw_signal contains 288121 bits == 490 frames of 588 bits each, plus a single extra bit so we can see the changes.

raw_signal = map(int, raw_signal)

# The 'delta' signal has one less entry:

delta_signal = [raw_signal[i] ^ raw_signal[i - 1] for i in range(1, len(raw_signal))]

if True:

    im = Image.new('1', (588, 490), 'red')

    for y in range(490):
        for x in range(588):
            if raw_signal[588 * y + x +1]:
                im.putpixel((x, y), 1)
            else:
                im.putpixel((x, y), 0)

    im.save("raw-signal.png")

if True:

    im = Image.new('1', (588, 490))

    for y in range(490):
        for x in range(588):
            if raw_signal[588 * y + x + 1]:
                im.putpixel((x, y), 0)
            else:
                im.putpixel((x, y), 1)

    im.save("raw-signal-inverted.png")

if True:

    im = Image.new('1', (588, 490))

    for y in range(490):
        for x in range(588):
            if delta_signal[588 * y + x]:
                im.putpixel((x, y), 0)
            else:
                im.putpixel((x, y), 1)

    im.save("delta-signal.png")

if True:

    im = Image.new('RGB', (588, 490))

    colormap = {
        0 : (255, 204, 255), # subchannel byte
        1 : (204, 255, 255), # AUDIO byte (MSB) LEFT
        2 : (204, 204, 255), # AUDIO byte (LSB) LEFT
        3 : (204, 255, 255), # AUDIO byte (MSB) LEFT
        4 : (204, 204, 255), # AUDIO byte (LSB) LEFT
        5 : (204, 255, 255), # AUDIO byte (MSB) LEFT
        6 : (204, 204, 255), # AUDIO byte (LSB) LEFT
        7 : (204, 255, 255), # AUDIO byte (MSB) RIGHT
        8 : (204, 204, 255), # AUDIO byte (LSB) RIGHT
        9 : (204, 255, 255), # AUDIO byte (MSB) RIGHT
        10: (204, 204, 255), # AUDIO byte (LSB) RIGHT
        11: (204, 255, 255), # AUDIO byte (MSB) RIGHT
        12: (204, 204, 255), # AUDIO byte (LSB) RIGHT
        13: (204, 255, 204), # error correction (C2 a.k.a. Q)
        14: (204, 255, 204), # error correction (C2 a.k.a. Q)
        15: (204, 255, 204), # error correction (C2 a.k.a. Q)
        16: (204, 255, 204), # error correction (C2 a.k.a. Q)
        17: (204, 255, 255), # AUDIO byte (MSB) LEFT
        18: (204, 204, 255), # AUDIO byte (LSB) LEFT
        19: (204, 255, 255), # AUDIO byte (MSB) LEFT
        20: (204, 204, 255), # AUDIO byte (LSB) LEFT
        21: (204, 255, 255), # AUDIO byte (MSB) LEFT
        22: (204, 204, 255), # AUDIO byte (LSB) LEFT
        23: (204, 255, 255), # AUDIO byte (MSB) RIGHT
        24: (204, 204, 255), # AUDIO byte (LSB) RIGHT
        25: (204, 255, 255), # AUDIO byte (MSB) RIGHT
        26: (204, 204, 255), # AUDIO byte (LSB) RIGHT
        27: (204, 255, 255), # AUDIO byte (MSB) RIGHT
        28: (204, 204, 255), # AUDIO byte (LSB) RIGHT
        29: (255, 255, 204), # error correction (C1 a.k.a. P)
        30: (255, 255, 204), # error correction (C1 a.k.a. P)
        31: (255, 255, 204), # error correction (C1 a.k.a. P)
        32: (255, 255, 204)  # error correction (C1 a.k.a. P)
    }

    for y in range(490):
        for x in range(588):
            if delta_signal[588 * y + x]:
                im.putpixel((x, y), (0, 0, 0))
            else:
                if x < 24:
                    im.putpixel((x, y), (255, 204, 204)) # sync pattern
                elif (x - 24) % 17 < 3:
                    im.putpixel((x, y), (204, 204, 204)) # merge bits
                else:
                    im.putpixel((x, y), colormap[(x - 24) // 17])

    im.save("delta-signal-colored.png")
